core/app/models/spree/payment/processing.rb

```diff
@@ -1,11 +1,36 @@
 module Spree
   class Payment < Spree::Base
     module Processing
+      # "process!" means:
+      #   - Do nothing when:
+      #     - There is no payment method
+      #     - The payment method does not require a source
+      #     - The payment is in the "processing" state
+      #     - 'auto_capture?' is false and the payment is already authorized.
+      #   - Raise an exception when:
+      #     - The source is missing or invalid
+      #     - The payment is in a state that cannot transition to 'processing'
+      #       (failed/void/invalid states). Note: 'completed' can transition to
+      #       'processing' and thus calling #process! on a completed Payment
+      #       will attempt to re-authorize/re-purchase the payment.
+      #   - Otherwise:
+      #     - If 'auto_capture?' is true:
+      #       - Call #purchase on the payment gateway. (i.e. authorize+capture)
+      #         even if the payment is already failed/completed/etc.
+      #     - Else:
+      #       - Call #authorize on the payment gateway even if the payment is
+      #         already authorized/failed/completed/etc.
       def process!
-        if payment_method && payment_method.auto_capture?
+        return if payment_method.nil?
+
+        if payment_method.auto_capture?
           purchase!
         else
-          authorize!
+          if pending?
```

Comment:
I believe you are correct -- in that case I think a credit card (at least Braintree provider) would re-auth, resulting in too over authorization.  I *think* a store credit would actually end up working just fine (not that that was by design...).  I'm not sure what the right thing to do there is though since we don't have a `#capture!` method that works in the same way as the `#purchase!` and `#authorize!` methods.  What do you think of opening a separate issue to address that?  This store credit auth issue is a bug that we really need to get fixed.